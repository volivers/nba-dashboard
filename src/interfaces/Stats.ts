import { Game } from "./Game";
import { Player } from "./Player";
import { Team } from "./Team";

export interface Stats {
  ast: number;
  blk: number;
  dreb: number;
  fg3Pct: number;
  fg3a: number;
  fg3m: number;
  fgPct: number;
  fga: number;
  fgm: number;
  ftPct: number;
  fta: number;
  ftm: number;
  game: Game;
  id: number;
  min: string;
  oreb: number;
  pf: number;
  player: Player;
  pts: number;
  reb: number;
  stl: number;
  team: Team;
  turnover: number;
}
