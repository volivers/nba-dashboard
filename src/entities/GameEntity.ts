import { Game } from '../interfaces/Game';
import { TeamEntity, deserializeTeamEntity, mockedAwayTeamEntity, mockedHomeTeamEntity } from './TeamEntity';

export interface GameEntity {
  date: string;
  home_team: TeamEntity;
  home_team_score: number;
  id: number;
  period: number;
  postseason: boolean;
  season: number;
  status: string;
  time: string;
  visitor_team: TeamEntity;
  visitor_team_score: number;
}

export const deserializeGameEntity = (entity: GameEntity): Game => ({
  date: entity.date,
  homeTeam: deserializeTeamEntity(entity.home_team),
  homeTeamScore: entity.home_team_score,
  id: entity.id,
  period: entity.period,
  postseason: entity.postseason,
  season: entity.season,
  status: entity.status,
  time: entity.time,
  visitorTeam: deserializeTeamEntity(entity.visitor_team),
  visitorTeamScore: entity.visitor_team_score,
});

export const mockedGameEntity: GameEntity = {
  date: '2000-11-01T00:00:00.000Z',
  home_team: mockedHomeTeamEntity,
  home_team_score: 100,
  id: 1,
  period: 4,
  postseason: true,
  season: 1,
  status: 'Final',
  time: '',
  visitor_team: mockedAwayTeamEntity,
  visitor_team_score: 80,
};
