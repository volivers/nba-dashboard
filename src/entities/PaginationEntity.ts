import { Pagination } from '../interfaces/Pagination';

export interface PaginationEntity {
  current_page: number;
  next_page: number;
  per_page: number;
}

export const deserializePaginationEntity = (entity: PaginationEntity): Pagination => ({
  currentPage: entity.current_page,
  nextPage: entity.next_page,
  perPage: entity.per_page,
});

export const mockedPaginationEntity: PaginationEntity = {
  current_page: 1,
  next_page: 2,
  per_page: 25,
};
