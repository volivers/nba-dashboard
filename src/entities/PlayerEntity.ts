import { Player } from '../interfaces/Player';
import { TeamEntity, deserializeTeamEntity, mockedHomeTeamEntity } from './TeamEntity';

export interface PlayerEntity {
  first_name: string;
  height_feet?: number;
  height_inches?: number;
  id: number;
  last_name: string;
  position: string;
  team: TeamEntity;
  weight_pounds?: number;
}

export const deserializePlayerEntity = (entity: PlayerEntity): Player => ({
  firstName: entity.first_name,
  heightFeet: entity?.height_feet,
  heightInches: entity?.height_inches,
  id: entity.id,
  lastName: entity.last_name,
  position: entity.position,
  team: deserializeTeamEntity(entity.team),
  weightPounds: entity?.weight_pounds,
});

export const mockedPlayerEntity: PlayerEntity = {
  first_name: 'Neemias',
  id: 88,
  last_name: 'Queta',
  position: 'C',
  team: mockedHomeTeamEntity,
};
