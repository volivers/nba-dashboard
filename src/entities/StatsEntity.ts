import { Stats } from '../interfaces/Stats';
import { GameEntity, deserializeGameEntity, mockedGameEntity } from './GameEntity';
import { PlayerEntity, deserializePlayerEntity, mockedPlayerEntity } from './PlayerEntity';
import { TeamEntity, deserializeTeamEntity, mockedHomeTeamEntity } from './TeamEntity';

export interface StatsEntity {
  ast: number;
  blk: number;
  dreb: number;
  fg3_pct: number;
  fg3a: number;
  fg3m: number;
  fg_pct: number;
  fga: number;
  fgm: number;
  ft_pct: number;
  fta: number;
  ftm: number;
  game: GameEntity;
  id: number;
  min: string;
  oreb: number;
  pf: number;
  player: PlayerEntity;
  pts: number;
  reb: number;
  stl: number;
  team: TeamEntity;
  turnover: number;
}

export const deserializeStatsEntity = (entity: StatsEntity): Stats => ({
  ast: entity.ast,
  blk: entity.blk,
  dreb: entity.dreb,
  fg3Pct: entity.fg3_pct,
  fg3a: entity.fg3a,
  fg3m: entity.fg3m,
  fgPct: entity.fg_pct,
  fga: entity.fga,
  fgm: entity.fgm,
  ftPct: entity.ft_pct,
  fta: entity.fta,
  ftm: entity.ftm,
  game: deserializeGameEntity(entity.game),
  id: entity.id,
  min: entity.min,
  oreb: entity.oreb,
  pf: entity.pf,
  player: deserializePlayerEntity(entity.player),
  pts: entity.pts,
  reb: entity.reb,
  stl: entity.stl,
  team: deserializeTeamEntity(entity.team),
  turnover: entity.turnover,
});

export const mockedStatsEntity: StatsEntity = {
  ast: 10,
  blk: 4,
  dreb: 15,
  fg3_pct: 1,
  fg3a: 10,
  fg3m: 10,
  fg_pct: 10,
  fga: 10,
  fgm: 10,
  ft_pct: 10,
  fta: 10,
  ftm: 10,
  game: mockedGameEntity,
  id: 1,
  min: '18',
  oreb: 5,
  pf: 14,
  player: mockedPlayerEntity,
  pts: 20,
  reb: 10,
  stl: 5,
  team: mockedHomeTeamEntity,
  turnover: 10,
};
