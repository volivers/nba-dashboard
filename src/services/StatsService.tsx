import axios from 'axios';
import { PaginationEntity, deserializePaginationEntity } from '../entities/PaginationEntity';
import { Pagination } from '../interfaces/Pagination';
import { Params, serializeParams } from '../interfaces/Params';
import { StatsEntity, deserializeStatsEntity } from '../entities/StatsEntity';
import { Stats } from '../interfaces/Stats';

export interface StatsResponse {
  data: Stats[];
  meta: Pagination;
}

export interface StatsResponseEntity {
  data: StatsEntity[];
  meta: PaginationEntity;
}

export const deserializeStatsResponse = (entity: StatsResponseEntity): StatsResponse => ({
  data: entity.data.map(deserializeStatsEntity),
  meta: deserializePaginationEntity(entity.meta)
});

export const STATS_API_URL = process.env.REACT_APP_API_URL + '/stats';

export class StatsService {
  public static async getAll(params?: Params): Promise<StatsResponse> {
    const response = await axios<StatsResponseEntity>({
      method: 'GET',
      url: STATS_API_URL,
      params: { ...serializeParams(params) },
      headers: {
        'Content-Type': 'application/json',
        'X-RapidAPI-Key': process.env.REACT_APP_API_KEY,
        'X-RapidAPI-Host': 'free-nba.p.rapidapi.com'
      }
    });

    return deserializeStatsResponse(response.data);
  };
};
