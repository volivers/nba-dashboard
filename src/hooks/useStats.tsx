import { useQuery } from 'react-query';
import { Params } from '../interfaces/Params';
import { StatsService } from '../services/StatsService';

export const useStats = (params?: Params) => {
  return useQuery(['stats', params], () => StatsService.getAll(params), {
    keepPreviousData: true
  });
};
