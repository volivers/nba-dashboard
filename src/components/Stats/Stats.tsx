import React, { useState } from 'react';
import { useStats } from '../../hooks/useStats';
import PaginatedTable from '../Table/PaginatedTable';
import { Player } from '../../interfaces/Player';
import { formatDate } from '../../utils/formatDate';

export interface StatsProps {
  activeGameId?: number;
  description: string;
  selectedPlayer?: Player;
  title: string;
}

const Stats = ({ activeGameId, description, selectedPlayer, title }: StatsProps) => {
  const [currentPage, setCurrentPage] = useState(1);

  const { data: results, isLoading } = useStats({
    ...selectedPlayer && { playerIds: [selectedPlayer.id] },
    ...activeGameId && { gameIds: [activeGameId] },
    page: currentPage,
    perPage: 10,
  });

  if (isLoading) return <p>Loading stats...</p>;

  return (
    <PaginatedTable
      description={description}
      isLastPage={!results?.meta.nextPage || results?.meta.nextPage === results?.meta.currentPage}
      onPageChange={setCurrentPage}
      page={currentPage}
      title={title}
    >
      <thead>
        <tr>
          <th />
          <th>Ast</th>
          <th>Blk</th>
          <th>Pts</th>
          <th>Reb</th>
          <th>Stl</th>
          <th>Turnover</th>
          <th>Min</th>
        </tr>
      </thead>
      <tbody>
        {results?.data.map(({ ast, blk, game, id, min, player, pts, reb, stl, turnover }) => (
          min && (
            <tr key={id}>
              <td className="text-nowrap">
                {activeGameId ? `${player.firstName} ${player.lastName}` : formatDate(game.date)}
              </td>
              <td>{ast}</td>
              <td>{blk}</td>
              <td>{pts}</td>
              <td>{reb}</td>
              <td>{stl}</td>
              <td>{turnover}</td>
              <td>{min}</td>
            </tr>
          )
        ))}
      </tbody>
    </PaginatedTable >
  );
};

export default Stats;
