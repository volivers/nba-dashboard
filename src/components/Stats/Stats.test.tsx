import { render, screen } from '@testing-library/react';
import Stats, { StatsProps } from './Stats';
import { useStats } from '../../hooks/useStats';
import { deserializePaginationEntity, mockedPaginationEntity } from '../../entities/PaginationEntity';
import { StatsResponse } from '../../services/StatsService';
import { deserializeStatsEntity, mockedStatsEntity } from '../../entities/StatsEntity';
import { deserializePlayerEntity, mockedPlayerEntity } from '../../entities/PlayerEntity';

jest.mock('../../hooks/useStats');
const mockedUseStats = useStats as jest.Mock;

describe('Stats', () => {
  let props: StatsProps;
  let mockedStatsResponse: StatsResponse;

  beforeEach(() => {
    props = {
      description: 'Lorem ipsum',
      selectedPlayer: deserializePlayerEntity(mockedPlayerEntity),
      title: 'Stats',
    };

    mockedStatsResponse = {
      data: [deserializeStatsEntity(mockedStatsEntity)],
      meta: deserializePaginationEntity(mockedPaginationEntity),
    };

    mockedUseStats.mockResolvedValue({
      ...mockedStatsResponse,
      isLoading: false,
    });
  });

  it('renders properly', () => {
    const { container } = render(<Stats {...props} />);

    expect(container).toMatchSnapshot();
  });

  describe('when Stats are loading', () => {
    beforeEach(() => {
      mockedUseStats.mockReturnValue({
        ...mockedStatsResponse,
        isLoading: true,
      });
    });

    it('renders a loading message', () => {
      render(<Stats {...props} />);

      expect(screen.getByText(/Loading stats.../)).toBeInTheDocument();
    });
  });
});
