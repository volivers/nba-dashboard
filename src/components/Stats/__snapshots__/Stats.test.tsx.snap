// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stats renders properly 1`] = `
<div>
  <div>
    <h1
      class="my-3"
    >
      Stats
    </h1>
    <div
      class="d-flex justify-content-between"
    >
      <p>
        <small>
          Lorem ipsum
        </small>
      </p>
      <ul
        class="mt-3 pagination"
      >
        <li
          class="page-item disabled"
        >
          <span
            class="page-link"
          >
            <span
              aria-hidden="true"
            >
              ‹
            </span>
            <span
              class="visually-hidden"
            >
              Previous
            </span>
          </span>
        </li>
        <li
          class="page-item disabled"
        >
          <span
            class="page-link"
          >
            <span
              aria-hidden="true"
            >
              ›
            </span>
            <span
              class="visually-hidden"
            >
              Next
            </span>
          </span>
        </li>
      </ul>
    </div>
    <div
      class="table-responsive"
    >
      <table
        class="table table-hover"
      >
        <thead>
          <tr>
            <th />
            <th>
              Ast
            </th>
            <th>
              Blk
            </th>
            <th>
              Pts
            </th>
            <th>
              Reb
            </th>
            <th>
              Stl
            </th>
            <th>
              Turnover
            </th>
            <th>
              Min
            </th>
          </tr>
        </thead>
        <tbody />
      </table>
    </div>
  </div>
</div>
`;
